bin/common/utils.js

```diff
@@ -43,17 +43,24 @@ exports.loadNuxtConfig = (argv) => {
   options.mode =
     (argv.spa && 'spa') || (argv.universal && 'universal') || options.mode
 
-  return options
-}
+  // Server options
+  if (!options.server) {
+    options.server = {}
+  }
 
-exports.getLatestHost = (argv) => {
-  const port =
-    argv.port || process.env.NUXT_PORT || process.env.PORT || process.env.npm_package_config_nuxt_port
-  const host =
+  options.server.port =
```

Comment:
This part is better to move into lib/common/nuxt.config.js, default value is from env.

```js
server: {
    host: process.env.NUXT_HOST ||
      process.env.HOST ||
      process.env.HOST ||
      process.env.npm_package_config_nuxt_host,
    port: argv.port ||
      process.env.NUXT_PORT ||
      process.env.PORT ||
      process.env.npm_package_config_nuxt_port
  }
```

So the user defined server.host or port will have higher priority.

And also arguments from command line has highest priority, so only keep `server.host = argv.host`(also port) when they are not undefined in loadeNuxtConfig.

Reply:
Let me make a check, so the priority is: `argv > nuxt.config.js > env`, right?

Reply:
I'd suggest `argv > NUXT_PORT env > PORT env > nuxt.config.js > package.json`

Reply:
@manniL Yes, that's what I was attempting to do at the beginning. Now I get a little bit confused. 😅

Reply:
@antfu You can move the `port`/`host` assignments from `bin/common/utils.js` to `lib/common/nuxt.config.js` because `process.env` is available there. So the "default values" are all in one place :relaxed:

Reply:
@manniL If the `process.env` is default value set in `lib/common/nuxt.config.js`, then user defined values will override the env ones. In other words, the `nuxt.config.js` has higher priority than `env`. I am not sure if my understanding is correct.

Reply:
I see the problem though. if you copy it like this the order is `argv > NUXT_XYZ env > XYZ env > package.json > nuxt.config.js`

If you insert this in `nuxt.config.js`, the nuxt config value will be the one with the **least** priority. (cc @clarkdo)

```js
server: {
    host: process.env.NUXT_HOST ||
      process.env.HOST ||
      process.env.npm_package_config_nuxt_host ||
	  'localhost',
    port: argv.port ||
      process.env.NUXT_PORT ||
      process.env.PORT ||
      process.env.npm_package_config_nuxt_port ||
      3000
  }
}
```

Reply:
For priority order: `argv > NUXT_PORT env > PORT env > nuxt.config.js > package.json`

```js
exports.loadNuxtConfig = (argv) => {

  /* other stuff... */

  // Server options
  options.server.port =
    argv.port ||
    process.env.NUXT_PORT ||
    process.env.PORT ||
    options.server.port ||
    process.env.npm_package_config_nuxt_port

  options.server.host =
    argv.hostname ||
    process.env.NUXT_HOST ||
    process.env.HOST ||
    options.server.host ||
    process.env.npm_package_config_nuxt_host

  return options
}
```
```js
// lib/common/nuxt.config.js
server: {
  port: undefined,
  host: undefined,
}
```
and leave `lib/common/nuxt.config.js` undefined, otherwise the `package.json` will never be applied.

Reply:
In my opinion, priority only has one scenario:  argv > nuxt.config.

Also the env is the default value of nuxt.config.js, if you don't set `server.host`, it works like: argv > nuxt.config = env.

The order should be related to the level of project and system, env is depend on the system, config is depend on the projectm and argv is depend on specific build or start command, so the order should be argv > config > env.

Reply:
And the overall default value is defined in
https://github.com/nuxt/nuxt.js/blob/04e53b0f9f50d2fc95de046b55f68147b617ff8e/lib/core/nuxt.js#L119-L124

Reply:
@clarkdo The problem is: if you've set the host in the `package.json` and in the `nuxt.config.js`, `package.json` values would have a higher priority because they'll be merged into the `env`, which isn't intuitive at all :thinking:

Reply:
I am not a heavy linux user. In Windows, most of time I use `cross-env PORT=8888 npm run dev` to specify the port temporary (some project might not have argv available). So I suppose the env have higher priority than config. This is from my little developing experience so it may not be correct.

Reply:
@antfu This is completely right and fulfilled currently (and also in your proposals):

`const host = process.env.NUXT_HOST || ... || '0.0.0.0'` means that if env is set, it'll be used instead of the default `'0.0.0.0'`

Reply:
@manniL 
If set host both in package.json and nuxt.config, package.json shouldn't have higher priority than nuxt.config, because `nuxt.config.js` is closer to application.

Reply:
I am kinda confuse which approach should I change to 😅

Reply:
@clarkdo That's what I am saying :see_no_evil:  So we **cannot** move the declaration into `lib/common/nuxt.config.js`:

```js

server: {
    host: argv.port || //argument
      process.env.NUXT_HOST || //env, nuxt
      process.env.HOST || //env
      process.env.npm_package_config_nuxt_host || // package.json <- PROBLEM
	  'localhost', // default from config
    port: argv.port || //argument
      process.env.NUXT_PORT || //env, nuxt
      process.env.PORT || //env
      process.env.npm_package_config_nuxt_port || //package.json <- PROBLEM
      3000 //default.
  }
```

https://github.com/nuxt/nuxt.js/pull/3701#discussion_r209452056 seems alright but should be moved to `lib/common/options.js` for better separation

Reply:
Let's take an example:

```
// package.json
nuxt: {
  host: localhost
}

// nuxt.config.js
server: {
  host: '127.0.0.1'
}

// lib/common/nuxt.config.js
server: {
    host: argv.port || //argument
      process.env.NUXT_HOST || //env, nuxt
      process.env.HOST || //env
      process.env.npm_package_config_nuxt_host || // package.json <- PROBLEM
	  'localhost', // default from config
}

```


The final `server.host` will be `127.0.0.1` because config in nuxt.config.js will override the default value, I think it's correct

Reply:
hostname should have highest priority, like I said before:

> And also arguments from command line has highest priority, so only keep server.host = argv.host(also port) when they are not undefined in loadeNuxtConfig.

But I doubt if NUXT_HOST should have higher priority than `server.host`, the thought of mine is that if we want to have a new config `server`, we should merge all host related config into one entry.

Reply:
@antfu Good point!

@clarkdo I think we both want the same :see_no_evil: 

Let's step through the example again (please bear with me :sweat_smile:)

Say we have the port resolved to `const post =argv.port || options.server.port`.
The default for `options.server.port` is:

```js
{
  port:  process.env.NUXT_PORT ||
         process.env.PORT ||
         process.env.npm_package_config_nuxt_port ||
         3000
}
```

If we now set the port through `nuxt.config.js`, the `env` variables (`NUXT_PORT`, `PORT`) won't be taken into account at all and are overridden with the value from `nuxt.config.js`

Reply:
Yes, that's correct.

Reply:
@antfu
> so only keep server.host = argv.host(also port) when they are not undefined in loadeNuxtConfig.

Reply:
@clarkdo So you'd prioritize `nuxt.config.js` over `NUXT_PORT/HOST`? 
I'd see environment variables closer to the program but maybe that's only my perception. :man_shrugging:

Reply:
@clarkdo I am fine if you think `config.js` should over `env`, I'll change it. But I still do not think `env.PORT` is system level (which is also meaningless to specify a PORT to system). `env.PORT` is more command line flexible than the file `nuxt.config.js`. So I'd prefer `env > config`.

Reply:
Hmm, it also makes sense, maybe let @Atinux decide.